### PR TITLE
Fix bug on the backpropagation of LayerNorm when create_graph=True

### PR DESCRIPTION
**Pull request by vinnamkim**
Solve an issue #41332

I found the bug at #41332 is caused by LayerNorm.

Current implementations of LayerNorm have a disparity between 
1. [`create_graph=False` CUDA implementation](https://github.com/BIT-silence/pytorch/blob/dde3d5f4a8f713ecc4649d776565b68ca75ae5c8/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L145)
2. [`create_graph=True` implementation](https://github.com/BIT-silence/pytorch/blob/dde3d5f4a8f713ecc4649d776565b68ca75ae5c8/tools/autograd/templates/Functions.cpp#L2536)

With this bug-fix, #41332 is solved.

@Ailing @BIT-silence

Signed-off-by: Vinnam Kim <vinnamkim@gmail.com>

### CODE REVIEW

**Comment by BIT-silence** [comment]
Thanks for fixing this.

**Review by xiaomengy — APPROVED** [review]

**Review by albanD — CHANGES_REQUESTED** [review]
Thanks for the fix.

I am surprised that our tests didn't caught this. The layer norm double backward should be checked with finite difference. It is properly listed in https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_nn.py
Or is it that we don't test for cases where gamma is different from 1?

Also the ci failures looks related.

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -2553,7 +2553,7 @@ infinitely_differentiable_native_layer_norm_backward(
       const Tensor& a = rstd_tensor;
       const Tensor b = (db * mean_tensor - ds) * rstd_cube * s;
       const Tensor c = -b * mean_tensor - db * rstd_tensor * s;
-      dX = a * dY_tensor + b * X_tensor + c;
+      dX = a * dY_tensor * gamma_tensor + b * X_tensor + c;
```
Comment (BIT-silence) [inline tools/autograd/templates/Functions.cpp:2556]:
```suggestion
      if (gamma.defined()) {
        dX = a * dY_tensor * gamma_tensor + b * X_tensor + c;
      } else {
        dX = a * dY_tensor + b * X_tensor + c;
      }
```

**Review by xiaomengy — CHANGES_REQUESTED** [review]

**Comment by BIT-silence** [comment]
> Thanks for the fix.
> 
> I am surprised that our tests didn't caught this. The layer norm double backward should be checked with finite difference. It is properly listed in https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_nn.py
> Or is it that we don't test for cases where gamma is different from 1?
> 
> Also the ci failures looks related.

I think somehow our double grad check didn't check the case when gamma and beta are defined. My comments in the code can help fix the test. We should check if gamma is defined before computing the grads.

**Comment by ngimel** [comment]
Please add a test in common_nn.py or in test_nn.py that would check this behavior (that is, would fail before your fix and pass with your fix).

**Comment by vinnamkim** [comment]
@albanD @BIT-silence @ngimel 

It is hard to write a test code using current test frameworks(checking double backward with finite difference). It actually passed double backward checks with simple test inputs, :<(maybe current implementation has analytically no effects on double backward?). So, I had no option but to add [ad-hoc test](https://github.com/pytorch/pytorch/pull/41595/commits/e8ec5b41d882d7eed3b6c9363f60e113841beebe#diff-d89baec73022f5f511c5beb5ce6498dfR8959).

**Comment by BIT-silence** [comment]
> @albanD @BIT-silence @ngimel
> 
> It is hard to write a test code using current test frameworks(checking double backward with finite difference). It actually passed double backward checks with simple test inputs, :<(maybe current implementation has analytically no effects on double backward?). So, I had no option but to add [ad-hoc test](https://github.com/pytorch/pytorch/pull/41595/commits/e8ec5b41d882d7eed3b6c9363f60e113841beebe#diff-d89baec73022f5f511c5beb5ce6498dfR8959).

I can understand it will be a little hard to add such a test in a short time. And when adding the test, there is also a risk that some other ops may fail this test. So I think we can open another issue to explain this missing test and tracking that in the new issue. What do you think? @ngimel

File: test/test_nn.py [hunk 2]
```diff
@@ -8956,6 +8956,37 @@ def __call__(self, input):
     input_size=(4, 16),
     fullname='AdaptiveLogSoftmax'))
 
+class _LayerNormWithNonDefaultWeight(nn.LayerNorm):
+    def __init__(self, normalized_shape, eps, elementwise_affine):
+        super().__init__(normalized_shape, eps, elementwise_affine)
+        self.weight.data = 0.1 * torch.ones_like(self.weight.data)
```
Comment (albanD) [inline test/test_nn.py:8962]: You should not use `.data` anymore.
Please use `with torch.no_grad():` here.

File: test/test_nn.py [hunk 3]
```diff
@@ -8956,6 +8956,37 @@ def __call__(self, input):
...
+class TestLayerNormWithNonDefaultWeight(NewModuleTest):
+    def __init__(self):
+        super().__init__(
+            constructor=lambda: _LayerNormWithNonDefaultWeight((16,), 1e-5, True),
+            input_size=(4, 4, 16),
+            check_gradgrad=True,
+            fullname='LayerNormWithNonDefaultWeight')
+
+    def compare_two_gradients(self, test_case, module, input):
+        num_threads = torch.get_num_threads()
```
Comment (albanD) [inline test/test_nn.py:8973]: Why is this needed? The accumulation operations are usually more precise when you have multiple threads actually.

File: test/test_nn.py [hunk 4]
```diff
@@ -8956,6 +8956,37 @@ def __call__(self, input):
...
+        rtol = 1e-3
+        test_case.assertTrue(torch.allclose(grads1, grads2, rtol, atol))
+
+        torch.set_num_threads(num_threads)
+
+    def _do_test(self, test_case, module, input):
+        self.compare_two_gradients(test_case, module, input)
+        super()._do_test(test_case, module, input)
+
+add_test(TestLayerNormWithNonDefaultWeight())
```
Comment (albanD) [inline test/test_nn.py:8989]: If you just want to add a simple test outside of the generic test framework, you can just add a new method to the `TestNN` class above and it will just run.

File: test/test_nn.py [hunk 5]
```diff
@@ -8956,6 +8956,37 @@ def __call__(self, input):
     input_size=(4, 16),
     fullname='AdaptiveLogSoftmax'))
 
+class _LayerNormWithNonDefaultWeight(nn.LayerNorm):
```
Comment (albanD) [inline test/test_nn.py:8959]: Why do you need a brand new class here? Can't you just update the weights from the original class after creation?

**Review by albanD — COMMENTED** [review]
Makes sense that you need a custom test. That looks good. We might want to open a task to make this a generic test indeed.

File: test/test_nn.py (hunk 2 above)
Comment (vinnamkim) [reply test/test_nn.py]: Fixed

File: test/test_nn.py (hunk 3 above)
Comment (vinnamkim) [reply test/test_nn.py]: Removed

File: test/test_nn.py (hunk 4 above)
Comment (vinnamkim) [reply test/test_nn.py]: Made a test class inherited from `TestCase`

File: test/test_nn.py (hunk 5 above)
Comment (vinnamkim) [reply test/test_nn.py]: Removed.

File: test/test_nn.py (hunk 4 above)
Comment (albanD) [reply test/test_nn.py]: We usually avoid new classes as, as you can see just below, we have custom setup code that is simpler to handle if you have a shared class.
Would it be possible to just move this test in the `TestNN` class that is just below?

File: test/test_nn.py [hunk 6]
```diff
@@ -309,7 +309,6 @@ def test_avg_pool3d_ceil_mode(self):
                 x.to('cuda'), ceil_mode=True, count_include_pad=True, kernel_size=(1, 2, 3), stride=2)
             self.assertTrue(not torch.isnan(y).any())
 
-
```
Comment (albanD) [inline test/test_nn.py:312]: nit: the linter might not be happy with the removal of this line.

**Review by albanD — APPROVED** [review]
That looks great thank!
Can you just rebase on top of master to make sure all CI is fine?

**Review by xiaomengy — APPROVED** [review]
The BC test should not be related.
